uwblueprint/focus-on-nature
# feat: Filter camps by year

## Notion ticket link

[Filter camps by year](https://www.notion.so/uwblueprintexecs/GET-camps-filtering-by-year-f226bcab30cc4f04b38e848ca7f153c8)



## Implementation description
* Updated the get all camps endpoint by accepting a `year` parameter which will filter the returned camps by the specified year
* Slack discussion for year filtering logic: https://uwblueprint.slack.com/archives/C02R9QK34UB/p1664037904985479
* TLDR; every camp has a list of camp sessions. We return any camp that has camp sessions in the specified year. If the camp has no camp sessions AKA the list is empty, we use the camp's creation date instead.
* Changes include updating camps model to include auto-generated createdAt field, updating route validators to account for a year being passed in, and filtering logic



## Steps to test
1. Use Postman and make a query to the endpoint like so: `http://localhost:5000/camp?campYear=2022`
2. Play around with the year passed and make sure we have proper error checking i.e can't pass invalid year or letters
3. Make sure the camps being returned contain ALL the camp sessions in the specified year and that no incorrect camp or camp session is being returned in the response
4. [Lower priority] Test the case If a camp has no camp sessions assigned to it but it was created in the year specified.

#### Sample test cases
`year`: 2022
- Camp created in 2022, campSession in 2022 -> return
- Camp created in 2022, no campSession in 2022 -> return
- Camp created in 2023, no campSession in 2021 -> don't return
- Camp created in 2023, campSession in 2022 -> X (should not exist)


## What should reviewers focus on?
* Filtering logic and possible errors
* Making sure the data being returned is correct


## Checklist
